### PR TITLE
Change metric.Producer to be an Option on Reader

### PR DESCRIPTION
**Pull request by dashpole**
Updates the MetricProducer implementation to comply with https://github.com/open-telemetry/opentelemetry-specification/pull/3613

### CODE REVIEW

**Comment by dashpole** [comment]
@MrAlias The register pattern came from your comment here: https://github.com/open-telemetry/opentelemetry-specification/pull/2722#discussion_r948113827

**Review by MrAlias — APPROVED** [review]

**Review by pellared — APPROVED** [review]

**Comment by pellared** [comment]
I tested (and analyzed for a a while as well) that after this PR is merged we can safely remove following fields from both readers:

```go
mu                sync.Mutex
isShutdown  bool
```

**Comment by dashpole** [comment]
I'm surprised this isn't caught by our concurrency tests...

**Comment by dashpole** [comment]
"fixed" our concurrency tests to expose the race condition.

File: sdk/metric/reader_test.go [hunk 1]
```diff
@@ -175,7 +172,7 @@ func (ts *readerTestSuite) TestMethodConcurrentSafe() {
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			_ = ts.Reader.Collect(ctx, nil)
+			_ = ts.Reader.Collect(ctx, &metricdata.ResourceMetrics{})
```
Comment (dashpole) [inline sdk/metric/reader_test.go:175]: This was necessary to demonstrate the data race condition.

**Comment by dashpole** [comment]
I believe switching back to atomic.Value to hold producers fixed the race condition.

**Review by MrAlias — APPROVED** [review]
:1st_place_medal:

**Comment by pellared** [comment]
> I believe switching back to atomic.Value to hold producers fixed the race condition.

👍 PS. I have not noticed that it was `atomic.Value` before 😬 Anyway, I am happy that I was able to find the issue 😄

**Pull request merged**